✨ Introduce `pixi` for project management

munechika-koyo opened the pull request:
# Summary
I am introducing the package management tool, `pixi`, to manage:
- package dependencies

- development tools (linter, formatter, etc.)

- task runners (building docs, testing, linting/formatting, etc.)

within each isolated virtual environment.

# Key changes:

- Added `pixi` manifest file for ~~project~~ workspace configuration.

- Included git-related files for ~~version control~~ pixi-related configure files.

- Updated `.gitignore` to accommodate `pixi` environments and lock files.

munechika-koyo committed:
1st commit for introducing `pixi`

Add pixi manifest file and git-related files which are automatically generated by `pixi`

Mateasek commented:
Hi @munechika-koyo,

I like pixi and now I'm in the middle of switching from uv to it, so forgive me if my questions/comments are out of the line.

What do we gain by adding pixi.toml and not integrating pixi into pyproject.toml?  Isn't the disadvantage of having pixi.toml the need to keep requirements in the two toml files synchronised?

 Since pipy is Cherab's main package source, wouldn't it be better to  prefer pipy packages in the toml file for now? 

Having only pyproject.toml and preferring pipy source, we can still keep the rest of the pixi features while not forcing users to migrate to pixi/conda. At least for some time.

munechika-koyo commented:
@Mateasek Thanks for your thoughtful question!

> What do we gain by adding pixi.toml and not integrating pixi into pyproject.toml? Isn't the disadvantage of having pixi.toml the need to keep requirements in the two toml files synchronised?

I used to manage a Python project similarly. However, because of integrating them into one toml file, it has relatively long lines, resulting in less readability. So I switched to this style, considering each file's role like:
- `pyproject.toml`: (1)Required Python project settings(`[build-systme]`, `[project]` tables), (2)python-related settings (build, packaging), (3)developing tool settings (lint/format, test, type checking, etc. `[tool.example]` table)
- `pixi.toml`: workspace settings like (1)dependencies required only when developing, (2)tasks (lint/format, build, doc, test, etc.), (3)packaging (with `pixi build` feature)

Currently, cherab's `pyproject.toml` is not so large that it might be feasible to unify those files.

> Since pipy is Cherab's main package source, wouldn't it be better to prefer pipy packages in the toml file for now?

The reason why I prefer using conda packages for `cherab`'s requirements is simply that I wanted to try the `pixi build` feature, and I presume it is better to use conda packages as many as possible for some reason.
Also, I think it should be fast to download conda packages using a [sharded system](https://prefix.dev/blog/sharded_repodata).

Although I believe this kind of thing should be okay for CI, we need to verify if the pypi package works well during deployment to pypi. `cibuildwheel` will neatly take care of it.


I think users, even those who want to install `cherab` from source, don't need `pixi`. Rather than `pixi`, it will empower the developer experience.

Mateasek commented:
Thank you for the explanation.

munechika-koyo committed:
Merge branch 'development' into feature/pixi

Mateasek reviewed (approved):
Hi @munechika-koyo ,

I would be happy to merge this if @jacklovell and @skuba31 agree. I think `Pixi` is the right step forward. Last year I tested Poetry, UV and Pixi and for me Pixi is the winner so far. As I said earlier, it significantly improves the packaging experience to a bearable level

munechika-koyo commented:
@Mateasek, Thank you for supporting this PR, and I apologize for not providing an example use case yet.
I'll work on it a little later.

munechika-koyo committed:
Update raysect dependency version to 0.9.1.*

munechika-koyo committed:
Revert "Update raysect dependency version to 0.9.1.*"

This reverts commit f2ee0d3754eb87477255b3ffef01dc37bcc17556.

munechika-koyo commented:
# Developer Tasks Guide

This document provides an overview of available development tasks configured in `pixi.toml` for the CHERAB project.
These tasks are organized by features and can be executed using the `pixi run <task-name>` command.
All tasks are displayed using `pixi task list` or `pixi task ls`.
It can be helpful to use help commands like  `pixi run help`, `pixi run <task-name> --help`, etc.

## Basic Tasks (Default Environment without `cherab`)

The default environment provides essential development utilities without building/installing `cherab` package:

### iPython terminal
```bash
pixi run ipython
```
-  **Purpose**: Launch `ipython` terminal. This task is available in all environments without `lint`.

### `clean`
```bash
pixi run clean
```
🔥 **Purpose**: Remove in-place build artifacts and temporary files (*.c, *.so, *.dylib)

This task recursively searches through the `cherab/` directory and removes compiled artifacts, helping maintain a clean development environment.

### Documentation Tasks

#### `doc-clean`
```bash
pixi run doc-clean
```
🔥 **Purpose**: Clean the docs build directory

Removes the entire `build` directory in the `docs/` folder to ensure fresh documentation builds.

#### `doc-serve`
```bash
pixi run doc-serve
```
🚀 **Purpose**: Start a local server for the docs

Starts an HTTP server on port 8000 serving the built documentation from `docs/build/html/`. Useful for previewing documentation changes locally.

## Feature-Based Tasks

### Test Feature

The test feature provides testing capabilities with the CHERAB package installed.

#### `test`
```bash
pixi run -e test test
```
🧪 **Purpose**: Run the test suite

Executes all unit tests in the CHERAB package using Python's unittest discovery mechanism with verbose output.

**Available test environments:**
- `test-py313`: Testing with Python 3.13
- `test-py39`: Testing with Python 3.9

### Documentation Feature

The docs feature provides comprehensive documentation building tools.

#### `doc-build`
```bash
pixi run -e docs doc-build
pixi run -e docs doc-build html  # Build HTML document (default)
```
📝 **Purpose**: Build the documentation

Uses Sphinx to build documentation in various formats (default: HTML). The target format can be specified as an argument.

**Dependencies:**
- **sphinx**: Documentation generator
- **sphinx_rtd_theme**: Read the Docs theme for consistent styling
- **sphinx-tabs**: Enables tabbed content in documentation

### Lint Feature

The lint feature provides a comprehensive suite of code quality and formatting tools.

#### Individual Linting Tasks

##### `ruff-check`
```bash
pixi run -e lint ruff-check
```
**Purpose**: Lint Python code with Ruff
- **ruff**: An extremely fast Python linter written in Rust that replaces multiple tools (flake8, isort, etc.)

##### `ruff-format`
```bash
pixi run -e lint ruff-format
```
**Purpose**: Format Python code with Ruff
- Automatically formats Python code according to configured style rules

##### `dprint`
```bash
pixi run -e lint dprint
```
**Purpose**: Format various file types with dprint
- **dprint**: A pluggable and configurable code formatter for multiple languages (JSON, TOML, Markdown, etc.)

##### `typos`
```bash
pixi run -e lint typos
```
**Purpose**: Fix typos automatically
- **typos**: A source code spell checker that can automatically fix common typos across the codebase

##### `taplo`
```bash
pixi run -e lint taplo
```
**Purpose**: Format TOML files
- **taplo**: A TOML toolkit providing formatting and validation for TOML configuration files

##### `actionlint`
```bash
pixi run -e lint actionlint
```
**Purpose**: Lint GitHub Actions workflows
- **actionlint**: Static checker for GitHub Actions workflow files, catching common mistakes and suggesting improvements

##### `cython-lint`
```bash
pixi run -e lint cython-lint
```
**Purpose**: Lint Cython files
- **cython-lint**: Specialized linter for Cython code (.pyx, .pxd files) ensuring code quality in performance-critical components

##### `blacken-docs`
```bash
pixi run -e lint blacken-docs
```
**Purpose**: Format Python code blocks in documentation
- **blacken-docs**: Applies Black formatting to Python code blocks within Markdown and reStructuredText files

##### `validate-pyproject`
```bash
pixi run -e lint validate-pyproject
```
**Purpose**: Validate pyproject.toml configuration
- **validate-pyproject**: Ensures the pyproject.toml file follows the correct schema and standards

#### Git Hook Management

##### `hooks`
```bash
pixi run -e lint hooks
```
🔗 **Purpose**: Install pre-commit hooks
- **lefthook**: Fast and powerful Git hooks manager that installs hooks to run linting automatically

##### `lefthook`
```bash
pixi run -e lint lefthook
```
🔗 **Purpose**: Run lefthook manually
- Executes the configured Git hooks without requiring a Git operation

##### `pre-commit`
```bash
pixi run -e lint pre-commit
```
🔗 **Purpose**: Run pre-commit checks
- Runs all configured pre-commit hooks in lefthook manually.

#### Comprehensive Linting

##### `lint`
```bash
pixi run -e lint lint
```
🧹 **Purpose**: Run all linters
- Executes all configured linting tools across the entire codebase, equivalent to running pre-commit hooks on all files

**Additional lint dependencies:**
- **shellcheck**: Shell script static analysis tool for bash/zsh scripts
- **validate-pyproject**: Validates Python project configuration files

## Usage Examples

### Daily Development Workflow

1. **Start development session:**
   ```bash
   pixi run clean  # Clean build artifacts
   ```

2. **Run tests:**
   ```bash
   pixi run test
   ```
	Choose the environment proposed in the terminal, or specify an environment:
	```bash
	pixi run -e test-py310 test
	```

3. **Lint and format code:**
   ```bash
   pixi run lint
   ```

4. **Build and serve documentation locally:**
   ```bash
   pixi run doc-build
   pixi run doc-serve
   ```

### Pre-commit Setup
> [!WARNING]  
> The lefthook configuration file (`.lefthook.yaml`) must be configured beforehand if you want to use this feature.
```bash
# Install Git hooks for automatic linting
pixi run hooks
```

### Documentation Development

```bash
# Clean previous builds
pixi run doc-clean

# Build HTML documentation
pixi run doc-build

# Serve locally for preview
pixi run doc-serve
# Visit http://localhost:8000
```

## Environment Selection

- **Default environment**: Basic development tools (only `ipython` currently )
- **Test environments**: `test`, `test-py313`, `test-py39`
- **Documentation environment**: `docs`
- **Linting environment**: `lint`

Use the `-e <environment>` flag to specify which environment to use for task execution if necessary.

Mateasek commented:
Simply amazing @munechika-koyo , thanks!

jacklovell commented on .gitignore:
```diff
@@ -21,4 +21,11 @@ build/
+
+# pixi environments
+.pixi/*
+!.pixi/config.toml
```
There's no `.pixi/config.toml` file in this PR. Does it really need to be un-ignored?

munechika-koyo replied:
These lines were automatically added by `pixi` when the `pixi init` command was run.
So it should probably be kept.

jacklovell commented on pixi.toml:
```diff
@@ -0,0 +1,153 @@
+[package.host-dependencies]
+uv = "*"
+python = "*"
+setuptools = "*"
+cython = ">=3.1"
+numpy = "*"
+raysect = "0.9.*"
```
Is there any way to auto-populate these from pyproject.toml? Otherwise we have 2 places to have to maintain dependency versions.

munechika-koyo replied:
There are two dependency sections:
- dependencies defined in pyproject.toml (or the `[pypi-dependencies]` table in pixi.toml), where packages are downloaded from the PyPI index.
- dependencies defined in `[dependencies]` in pixi.toml, where packages are downloaded by conda.

[Pixi prioritizes the conda packages](https://pixi.prefix.dev/latest/concepts/conda_pypi/), and only missing packages defined in pyproject.toml are downloaded through PyPI.
I prefer to use conda packages as much as possible to take advantage of conda-optimized packages, so I accounted for this maintenance cost.

munechika-koyo replied:
Additionally, in the `[package.***-dependencies]`, we must specify conda packages because they are required to build the `cherab` **conda** package (not for workspace packages).

jacklovell commented on pixi.toml:
```diff
@@ -0,0 +1,153 @@
+raysect = "0.9.*"
+
+[package.run-dependencies]
+scipy = "*"
+matplotlib-base = "*"
+pyopencl = "*"
+pocl = "*"
```
Ditto re:pyproject.toml

jacklovell commented on pixi.toml:
```diff
@@ -0,0 +1,153 @@
+compilers = ["c"]
+
+[workspace.build-variants]
+python = ["3.9", "3.10.*", "3.11.*", "3.12.*", "3.13.*"]
```
Can add 3.14 now too.

jacklovell commented on pixi.toml:
```diff
@@ -0,0 +1,153 @@
+    "'*.so'",
+    "-o",
+    "-name",
+    "'*.dylib'",
```
Add `*.html` here too, in case they've been produced by `cython --annotate` during development.

jacklovell commented on pixi.toml:
```diff
@@ -0,0 +1,153 @@
+    { arg = "target", default = "html" },
+], description = "📝 Build the docs" }
+
+# === Linting feature ===
```
A pylint linter would be useful too as it's very thorough. It's also however very slow, so perhaps worth separating it out from the others so `pixi run lint` is fast and `pixi run pylint` is thorough but slow.

Adding suitable configurations for each of the linters is for another pull request, after some TMC discussion I think.

jacklovell reviewed (commented):
This looks good. Can you add the instructions from the PR discussion into a markdown file in the `dev` directory, so they're easy to find for developers without having to come back to this PR?

munechika-koyo committed:
Update Python version range in pixi.toml

munechika-koyo committed:
Add HTML file removal to build artifacts cleanup

Added support for removing HTML build artifacts.

munechika-koyo committed:
Update Python version features in pixi.toml

Renaming python-featured environments to avoid specific Python versions

munechika-koyo committed:
Update channels and remove sphinx-tabs dependency

Updated channels to use 'conda-forge' instead of the prefix's URL.
Removed sphinx-tabs from pypi-dependencies.

munechika-koyo commented:
@jacklovell I added the `dev/pixi.md` file and modified `pixi.toml`.
I am so sorry for the late change, but I would appreciate it if you could review it again.

jacklovell reviewed (approved):
Looks good to me, thanks. @Mateasek since you also have some Pixi experience would you mind giving this one final check before merge?

Mateasek reviewed (approved):
I looked at the state of the PR now and I have to say I like what @munechika-koyo did. I don't have any change requests. I have to add that although I use pixi, I never used the features @munechika-koyo added and that the pixi.toml is more advanced than what I usually use. Since this step introduces pixi for development purposes, I think we can accept it as is, add it to the upcoming release and make improving changes if we discover some space for improvement. Now I don't see any. Thanks @munechika-koyo for working on it.
